Add option to ignore broken symlinks in directories

When using the "directory" src type, the plugin fails with the following error when it encounters a broken symlink:
`Caused by: java.io.FileNotFoundException: <filepath> (No such file or directory)
`
Most other maven plugins (for example the rpm-maven-plugin and maven-assembly-plugin) ignore any broken symlinks and process the rest of the directory.

When building large projects with lots of dependencies, there is a change one of them has a broken symlink somewhere, so plugins should be able to deal with it (by ignoring the broken symlink), or at least have an option to do that.

I can workaround the issue by first packaging the stuff I want into a tgz (using maven-assembly-plugin) and then using that as the src for jdeb, but jdeb should exhibit a behavior that's consistent with other maven plugins.

I've also attached my plugin configuration (project-specific and company-specific identifiers removed)
[jdeb.txt](https://github.com/user-attachments/files/15928821/jdeb.txt)

--- a/docs/maven.md
+++ b/docs/maven.md
@@ -121,6 +121,7 @@ classifier       | Artifact classifier                                          
 controlDir       | The directory containing the control files                                                 | No; defaults to `src/deb/control`
 installDir       | The default directory for the project artifact if no data section is present               | No; defaults to `/opt/${artifactId}`
 dataSet          | A list of directories, tarballs, or files to include in the deb package                    | No; defaults to include your maven artifact
+ignoreBrokenLinks| Indicates if broken symlinks should be ignored or cause build to fail                      | No; defaults to `false`
 changesIn        | The changes to add                                                                         | No
 changesOut       | The changes file generated                                                                 | No
 changesSave      | (NYI) The merged changes file                                                              | No

--- a/pom.xml
+++ b/pom.xml
@@ -19,7 +19,7 @@
   <artifactId>jdeb</artifactId>
   <packaging>maven-plugin</packaging>
   <name>jdeb</name>
-  <version>1.12</version>
+  <version>1.13-SNAPSHOT</version>
   <description>
     This library provides an Ant task and a Maven plugin to create Debian packages from Java builds in a truly cross
     platform manner. Build your Debian packages on any platform that has Java support. Windows, Linux, OS X - it doesn't

--- a/src/main/java/org/vafer/jdeb/DataBuilder.java
+++ b/src/main/java/org/vafer/jdeb/DataBuilder.java
@@ -285,7 +285,7 @@ class DataBuilder {
             // Get the offending file name from the exception to check
             // if it's a symlink.
             if (Files.isSymbolicLink(Paths.get(e.getFilePath())) && ignoreBrokenLinks) {
-                console.info("Ignoring broken symlink " + e.getMessage());
+                console.info("Ignoring broken symlink " + e.getFilePath());
             }
             else {
                 finishedWithoutErrors = false;

--- a/src/main/java/org/vafer/jdeb/DataBuilder.java
+++ b/src/main/java/org/vafer/jdeb/DataBuilder.java
@@ -26,11 +26,14 @@ import org.apache.commons.io.IOUtils;
 import org.vafer.jdeb.utils.Utils;
 
 import java.io.File;
+import java.io.FileNotFoundException;
 import java.io.FileOutputStream;
 import java.io.IOException;
 import java.io.InputStream;
 import java.math.BigInteger;
 import java.nio.ByteBuffer;
+import java.nio.file.Files;
+import java.nio.file.Paths;
 import java.security.DigestOutputStream;
 import java.security.MessageDigest;
 import java.security.NoSuchAlgorithmException;
@@ -84,12 +87,13 @@ class DataBuilder {
      * @param output
      * @param checksums
      * @param options Options used to build the data file
+     * @param ignoreBrokenLinks Whether to ignore broken symlinks in data directories
      * @return
      * @throws java.security.NoSuchAlgorithmException
      * @throws java.io.IOException
      * @throws org.apache.commons.compress.compressors.CompressorException
      */
-    BigInteger buildData(Collection<DataProducer> producers, File output, final StringBuilder checksums, TarOptions options) throws NoSuchAlgorithmException, IOException, CompressorException {
+    BigInteger buildData(Collection<DataProducer> producers, File output, final StringBuilder checksums, TarOptions options, boolean ignoreBrokenLinks) throws NoSuchAlgorithmException, IOException, CompressorException {
 
         final File dir = output.getParentFile();
         if (dir != null && (!dir.exists() || !dir.isDirectory())) {
@@ -277,6 +281,17 @@ class DataBuilder {
         try {
             for (DataProducer data : producers) {
                 data.produce(receiver);
+            }
+        } catch (FileNotFoundException e) {
+            // Get the offending file name from the exception message to check
+            // if it's a symlink.
+            String[] arr = e.getMessage().split(" ");
+            if (Files.isSymbolicLink(Paths.get(arr[0])) && ignoreBrokenLinks) {
+                console.info("Ignoring broken symlink " + arr[0]);
+            }
+            else {
+                finishedWithoutErrors = false;
+                throw e;
             }
         } catch (Exception e) {
             finishedWithoutErrors = false;

--- a/src/main/java/org/vafer/jdeb/DataBuilder.java
+++ b/src/main/java/org/vafer/jdeb/DataBuilder.java
@@ -26,7 +26,6 @@ import org.apache.commons.io.IOUtils;
 import org.vafer.jdeb.utils.Utils;
 
 import java.io.File;
-import java.io.FileNotFoundException;
 import java.io.FileOutputStream;
 import java.io.IOException;
 import java.io.InputStream;
@@ -282,12 +281,11 @@ class DataBuilder {
             for (DataProducer data : producers) {
                 data.produce(receiver);
             }
-        } catch (FileNotFoundException e) {
+        } catch (ProducerFileNotFoundException e) {
             // Get the offending file name from the exception message to check
             // if it's a symlink.
-            String[] arr = e.getMessage().split(" ");
-            if (Files.isSymbolicLink(Paths.get(arr[0])) && ignoreBrokenLinks) {
-                console.info("Ignoring broken symlink " + arr[0]);
+            if (Files.isSymbolicLink(Paths.get(e.getMessage())) && ignoreBrokenLinks) {
+                console.info("Ignoring broken symlink " + e.getMessage());
             }
             else {
                 finishedWithoutErrors = false;

--- a/src/main/java/org/vafer/jdeb/DataBuilder.java
+++ b/src/main/java/org/vafer/jdeb/DataBuilder.java
@@ -282,9 +282,9 @@ class DataBuilder {
                 data.produce(receiver);
             }
         } catch (ProducerFileNotFoundException e) {
-            // Get the offending file name from the exception message to check
+            // Get the offending file name from the exception to check
             // if it's a symlink.
-            if (Files.isSymbolicLink(Paths.get(e.getMessage())) && ignoreBrokenLinks) {
+            if (Files.isSymbolicLink(Paths.get(e.getFilePath())) && ignoreBrokenLinks) {
                 console.info("Ignoring broken symlink " + e.getMessage());
             }
             else {

--- a/src/main/java/org/vafer/jdeb/DebMaker.java
+++ b/src/main/java/org/vafer/jdeb/DebMaker.java
@@ -131,6 +131,9 @@ public class DebMaker {
     /** Defines the bigNumberMode of the tar file that is built */
     private String tarBigNumberMode;
 
+    /** Whether to ignore broken symlinks in DataBuilder */
+    private boolean ignoreBrokenLinks;
+
     private Long outputTimestampMs;
 
     private VariableResolver variableResolver;
@@ -257,6 +260,10 @@ public class DebMaker {
 
     public void setTarBigNumberMode(String tarBigNumberMode) {
         this.tarBigNumberMode = tarBigNumberMode;
+    }
+
+    public void setIgnoreBrokenLinks(boolean ignore) {
+        this.ignoreBrokenLinks = ignore;
     }
 
     public void setOutputTimestampMs(Long outputTimestampMs) {
@@ -487,7 +494,7 @@ public class DebMaker {
                 .compression(compression)
                 .longFileMode(tarLongFileMode)
                 .bigNumberMode(tarBigNumberMode);
-            BigInteger size = dataBuilder.buildData(dataProducers, tempData, md5s, options);
+            BigInteger size = dataBuilder.buildData(dataProducers, tempData, md5s, options, ignoreBrokenLinks);
 
             console.info("Building conffiles");
             List<String> tempConffiles = populateConffiles(conffilesProducers);

--- a/src/main/java/org/vafer/jdeb/ProducerFileNotFoundException.java
+++ b/src/main/java/org/vafer/jdeb/ProducerFileNotFoundException.java
@@ -4,6 +4,7 @@ import java.io.FileNotFoundException;
 
 public class ProducerFileNotFoundException extends FileNotFoundException {
 
+    private String filePath;
 
     public ProducerFileNotFoundException() {
         super();
@@ -16,5 +17,13 @@ public class ProducerFileNotFoundException extends FileNotFoundException {
     public ProducerFileNotFoundException(String message, Throwable cause) {
         super(message);
         this.initCause(cause);
+    }
+
+    public void setFilePath(String path) {
+        this.filePath = path;
+    }
+
+    public String getFilePath() {
+        return this.filePath;
     }
 }

--- a/src/main/java/org/vafer/jdeb/ProducerFileNotFoundException.java
+++ b/src/main/java/org/vafer/jdeb/ProducerFileNotFoundException.java
@@ -1,0 +1,18 @@
+package org.vafer.jdeb;
+
+import java.io.IOException;
+
+public class ProducerFileNotFoundException extends IOException {
+    
+    public ProducerFileNotFoundException() {
+        super();
+    }
+
+    public ProducerFileNotFoundException(String message) {
+        super(message);
+    }
+
+    public ProducerFileNotFoundException(String message, Throwable cause) {
+        super(message, cause);
+    }
+}

--- a/src/main/java/org/vafer/jdeb/ProducerFileNotFoundException.java
+++ b/src/main/java/org/vafer/jdeb/ProducerFileNotFoundException.java
@@ -1,9 +1,10 @@
 package org.vafer.jdeb;
 
-import java.io.IOException;
+import java.io.FileNotFoundException;
 
-public class ProducerFileNotFoundException extends IOException {
-    
+public class ProducerFileNotFoundException extends FileNotFoundException {
+
+
     public ProducerFileNotFoundException() {
         super();
     }
@@ -13,6 +14,7 @@ public class ProducerFileNotFoundException extends IOException {
     }
 
     public ProducerFileNotFoundException(String message, Throwable cause) {
-        super(message, cause);
+        super(message);
+        this.initCause(cause);
     }
 }

--- a/src/main/java/org/vafer/jdeb/maven/DebMojo.java
+++ b/src/main/java/org/vafer/jdeb/maven/DebMojo.java
@@ -353,6 +353,9 @@ public class DebMojo extends AbstractMojo {
     @Parameter(defaultValue = "gnu")
     private String tarBigNumberMode;
 
+    @Parameter(defaultValue = "false")
+    private boolean ignoreBrokenLinks;
+
     /**
      * Timestamp for reproducible output archive entries, either formatted as ISO 8601
      * <code>yyyy-MM-dd'T'HH:mm:ssXXX</code> or as an int representing seconds since the epoch (like
@@ -598,6 +601,7 @@ public class DebMojo extends AbstractMojo {
             debMaker.setDigest(digest);
             debMaker.setTarBigNumberMode(tarBigNumberMode);
             debMaker.setTarLongFileMode(tarLongFileMode);
+            debMaker.setIgnoreBrokenLinks(ignoreBrokenLinks);
             debMaker.setOutputTimestampMs(outputTimestampMs);
             debMaker.validate();
             debMaker.makeDeb();

--- a/src/main/java/org/vafer/jdeb/producers/AbstractDataProducer.java
+++ b/src/main/java/org/vafer/jdeb/producers/AbstractDataProducer.java
@@ -19,10 +19,12 @@ import org.apache.commons.compress.archivers.tar.TarArchiveEntry;
 import org.apache.tools.ant.types.selectors.SelectorUtils;
 import org.vafer.jdeb.DataConsumer;
 import org.vafer.jdeb.DataProducer;
+import org.vafer.jdeb.ProducerFileNotFoundException;
 import org.vafer.jdeb.mapping.Mapper;
 
 import java.io.File;
 import java.io.FileInputStream;
+import java.io.FileNotFoundException;
 import java.io.IOException;
 
 /**
@@ -85,7 +87,13 @@ public abstract class AbstractDataProducer implements DataProducer {
         TarArchiveEntry fileEntry = Producers.defaultFileEntryWithName(fileName);
         fileEntry.setSize(file.length());
         fileEntry = map(fileEntry);
-        Producers.produceInputStreamWithEntry(consumer, new FileInputStream(file), fileEntry);
+        try {
+            Producers.produceInputStreamWithEntry(consumer, new FileInputStream(file), fileEntry);
+        }
+        catch (FileNotFoundException e) {
+            ProducerFileNotFoundException pe = new ProducerFileNotFoundException(file.getAbsolutePath(), e);
+            throw pe;
+        }
     }
 
     public TarArchiveEntry map( final TarArchiveEntry pEntry ) {

--- a/src/main/java/org/vafer/jdeb/producers/AbstractDataProducer.java
+++ b/src/main/java/org/vafer/jdeb/producers/AbstractDataProducer.java
@@ -91,7 +91,8 @@ public abstract class AbstractDataProducer implements DataProducer {
             Producers.produceInputStreamWithEntry(consumer, new FileInputStream(file), fileEntry);
         }
         catch (FileNotFoundException e) {
-            ProducerFileNotFoundException pe = new ProducerFileNotFoundException(file.getAbsolutePath(), e);
+            ProducerFileNotFoundException pe = new ProducerFileNotFoundException(e.getMessage(), e);
+            pe.setFilePath(file.getAbsolutePath());
             throw pe;
         }
     }

--- a/src/test/java/org/vafer/jdeb/DataBuilderTestCase.java
+++ b/src/test/java/org/vafer/jdeb/DataBuilderTestCase.java
@@ -111,7 +111,7 @@ public final class DataBuilderTestCase extends Assert {
             builder.buildData(Arrays.asList(dirProducer), archive, new StringBuilder(), new TarOptions().compression(Compression.NONE), false);
             fail("buildData should not be successful with ignoreBrokenLinks set to false");
         }
-        catch (FileNotFoundException e) {
+        catch (ProducerFileNotFoundException e) {
             // ignoreBrokenLinks set to false, exception should be thrown.
         }
 

--- a/src/test/java/org/vafer/jdeb/DataBuilderTestCase.java
+++ b/src/test/java/org/vafer/jdeb/DataBuilderTestCase.java
@@ -18,7 +18,11 @@ package org.vafer.jdeb;
 
 import java.io.File;
 import java.io.FileInputStream;
+import java.io.FileNotFoundException;
 import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
 import java.util.Arrays;
 
 import org.apache.commons.compress.archivers.ArchiveEntry;
@@ -54,7 +58,7 @@ public final class DataBuilderTestCase extends Assert {
         fileset.setProject(project);
 
         StringBuilder md5s = new StringBuilder();
-        builder.buildData(Arrays.asList((DataProducer) new DataProducerFileSet(fileset)), new File("target/data.tar"), md5s, new TarOptions().compression(Compression.GZIP));
+        builder.buildData(Arrays.asList((DataProducer) new DataProducerFileSet(fileset)), new File("target/data.tar"), md5s, new TarOptions().compression(Compression.GZIP), false);
 
         assertTrue("empty md5 file", md5s.length() > 0);
         assertFalse("windows path separator found", md5s.indexOf("\\") != -1);
@@ -69,7 +73,7 @@ public final class DataBuilderTestCase extends Assert {
 
         DataProducer producer = new DataProducerFile(new File("pom.xml"), "/usr/share/myapp/pom.xml", null, null, null);
 
-        builder.buildData(Arrays.asList(producer), archive, new StringBuilder(), new TarOptions().compression(Compression.NONE));
+        builder.buildData(Arrays.asList(producer), archive, new StringBuilder(), new TarOptions().compression(Compression.NONE), false);
 
         int count = 0;
         try (TarArchiveInputStream in = new TarArchiveInputStream(new FileInputStream(archive))) {
@@ -90,9 +94,29 @@ public final class DataBuilderTestCase extends Assert {
         File archive = prepareArchive();
 
         DataBuilder builder = new DataBuilder(new NullConsole(), EXPECTED_MODIFIED_TIME);
-        builder.buildData(Arrays.asList(fileProducer, linkProducer, dirProducer), archive, new StringBuilder(), new TarOptions().compression(Compression.NONE));
+        builder.buildData(Arrays.asList(fileProducer, linkProducer, dirProducer), archive, new StringBuilder(), new TarOptions().compression(Compression.NONE), false);
 
         assertExpectedModTimeInArchive(archive);
+    }
+
+    @Test
+    public void testIgnoreBrokenLinks() throws Exception {
+        File dir = prepareSubdirWithSymlink();
+        DataProducer dirProducer = new DataProducerDirectory(dir, null, null, null);
+        File archive = prepareArchive();
+
+        DataBuilder builder = new DataBuilder(new NullConsole(), null);
+
+        try {
+            builder.buildData(Arrays.asList(dirProducer), archive, new StringBuilder(), new TarOptions().compression(Compression.NONE), false);
+            fail("buildData should not be successful with ignoreBrokenLinks set to false");
+        }
+        catch (FileNotFoundException e) {
+            // ignoreBrokenLinks set to false, exception should be thrown.
+        }
+
+        // ignoreBrokenLinks set to true, should not throw exception.
+        builder.buildData(Arrays.asList(dirProducer), archive, new StringBuilder(), new TarOptions().compression(Compression.NONE), true);
     }
 
     private File prepareArchive() {
@@ -115,6 +139,22 @@ public final class DataBuilderTestCase extends Assert {
 
         File file2 = new File(nestedDir, "file2.txt");
         file2.createNewFile();
+
+        return subDir;
+    }
+
+    private File prepareSubdirWithSymlink() throws IOException {
+        File subDir = new File("target/subDirWithLinks");
+        subDir.mkdir();
+
+        File file1 = new File(subDir, "file1.txt");
+        file1.createNewFile();
+
+        Path targetPath = Paths.get("nonExistentFile.txt");
+        Path linkPath = Paths.get(subDir.getPath() + "/link.txt");
+
+        // Create broken symlink in data directory
+        Files.createSymbolicLink(linkPath, targetPath);
 
         return subDir;
     }

